### PR TITLE
Add stack buffer overflows section

### PR DESCRIPTION
**Pull request by g-kouv**
I struggled with the level of detail for this section. In the end, I tried to include enough information on the topics I thought were relevant without getting into in-depth descriptions, to keep this manageable, while still including pointers for more information.

A lot of the links I've added did not seem suitable for the bibliography (no obvious author), so I've added them as inline links. I'm slightly worried that they might not be as obvious in the text as normal references. Let me know what you think!

Also note that various measures that would also be effective against stack smashing felt like they fit better in following sections, so I've indicated that in the text where relevant.

### CODE REVIEW

**Review by kbeyls — APPROVED** [review]
I like this!

As usual, I have many very minor nitpicks or suggestions. Feel free to disagree on them!

**Comment by kbeyls** [comment]
> I struggled with the level of detail for this section. In the end, I tried to include enough information on the topics I thought were relevant without getting into in-depth descriptions, to keep this manageable, while still including pointers for more information.

I thought the level of detail was right, so I don't think you need to worry about the result!

> A lot of the links I've added did not seem suitable for the bibliography (no obvious author), so I've added them as inline links. I'm slightly worried that they might not be as obvious in the text as normal references. Let me know what you think!

Maybe one way to make a distinction between using a reference and using a link could be: use a reference to point to a work that described or introduced something new or a new insight? And use links for things that give additional information, such as wikipedia articles or other articles explaining something in more detail? Just a thought...

If the links are not as obvious, we could use some style sheets for both the PDF and HTML output, e.g. adding underlines to indicate links and make them more obvious...

> Also note that various measures that would also be effective against stack smashing felt like they fit better in following sections, so I've indicated that in the text where relevant.

Makes sense to me!

File: book.md [hunk 1]
```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
...
+
+![Stack frame layout for stack buffer overflow example](img/stack){ width=60% }
+
+[^build-options]: The code is generated with the `-fno-stack-protector` option,
+  to ensure GCC's stack guard feature is disabled. We also used the `-O1`
+  optimization level.
+
+The exact details of the stack frame layout, including the ordering of
+variables and the exact control information stored, will depend on the specific
+compiler version we're using and the architecture we're compiling for.
```
[the remark this answers is not in the document]
Comment (g-kouv) [reply book.md:398]: Preferring the active voice over the passive voice appears to be considered good practice in technical writing. (And CMOS seems to suggest that it's usually preferable too: https://www.chicagomanualofstyle.org/book/ed17/part2/ch05/psec118.html). That said, I should probably use the present tense here ("we use" instead of "we're using"). I will also change "we" to "you".

[the remark this answers is not in the document]
Comment (kbeyls) [reply book.md:398]: that makes sense to me, thanks!

File: book.md [hunk 2]
```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
...
+In the code above, since the length of the argument is not checked before
+copying it into `dst`, we have a potential for a buffer overflow.
+
+When looking at code generated for AArch64 with GCC 11.2[^build-options],
+the stack layout looks like this:
+
+![Stack frame layout for stack buffer overflow example](img/stack){ width=60% }
+
+[^build-options]: The code is generated with the `-fno-stack-protector` option,
+  to ensure GCC's stack guard feature is disabled. We also used the `-O1`
```
[the remark this answers is not in the document]
Comment (JLouisKaplan-Arm) [reply book.md:393]: On second thoughts this is almost certainly not worth adding, but will leave up the comment as food for thought.

Also I now see you've mentioned GCC and clang later on in the section

File: book.md [hunk 3]
```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
...
+overwrite control data is going to overwrite the canary first, a corruption of
+the control data through a stack buffer overflow will be detected.
+
+This technique has a few limitations: first of all, it specifically aims to
+protect against stack buffer overflows, and does nothing to protect against
+stronger primitives (e.g. arbitrary write primitives). Control-flow integrity
+techniques, which are described in the next section, aim to protect the
+integrity of control data against any modification.
+
+Since stack canaries come with additional checks, compilers usually employ
```
[the remark this answers is not in the document]
Comment (JLouisKaplan-Arm) [reply book.md:439]: makes sense to me, I was wondering if it was worth making the explanation a little more explicit also. Not too sure, but something like might be handy:

> Since the use of stack canaries requires additional instructions for ensuring a canary's integrity, ...

File: book.md [hunk 4]
```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
...
+heuristics to determine which functions are considered vulnerable and therefore
+include the mitigation in their generated code. Since heuristics aren't always
```
Comment (JLouisKaplan-Arm) [inline book.md:441]: small nit/suggestion:

> ...heuristics to determine which functions are considered vulnerable and only include the mitigation in the generated code of those functions.

In the original wording, I think 'their' could be referring to the compiler's generated code (all of it) or the vulnerable functions' generated code, though I might just be overthinking it though.

**Review by JLouisKaplan-Arm — COMMENTED** [review]
Really nice! I liked it a lot. I added a few small comments / alternative wording suggestions but don't feel strongly about any of them. Mostly just thinking out loud.

One question I had that seemed more appropriate for here rather than inline: is there any plan for a section on heap overflow attacks? I thought of the question because I really liked the opening paragraph, and it almost seems like it could be the opening of a more general section that then has stack and heap overflow subsections. Happy to move this out to discussion/issue format if more appropriate.

Thanks!

File: book.md [hunk 5]
```diff
@@ -337,7 +337,153 @@ purposes of this chapter we will focus on:
...
+  to ensure GCC's stack guard feature is disabled. We also used the `-O1`
+  optimization level.
+
+The exact details of the stack frame layout, including the ordering of
+variables and the exact control information stored, will depend on the specific
+compiler version we're using and the architecture we're compiling for.
+
+As can be seen the stack diagram, an overflowing write can overwrite the saved
+frame pointer (FP) and link register (LR), which contains the return address,
+in `main`'s frame. This is an exploitable vulnerability.
```
[the remark this answers is not in the document]
Comment (g-kouv) [reply book.md:402]: I'll leave a TODO for this. Hopefully someone who is better at doing diagrams than me will want to pick this up! :)

File: book.md (hunk 4 above)
Comment (g-kouv) [reply book.md:441]: I think you are right. I'll rephrase this slightly to make this clearer, thanks!

File: book.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (g-kouv) [reply book.md:393]: I'll leave this as-is for now, then!

**Comment by g-kouv** [comment]
> One question I had that seemed more appropriate for here rather than inline: is there any plan for a section on heap overflow attacks? I thought of the question because I really liked the opening paragraph, and it almost seems like it could be the opening of a more general section that then has stack and heap overflow subsections. Happy to move this out to discussion/issue format if more appropriate.
>

I wasn't planning on writing a section specifically on heap overflows, but I'm still thinking about the format I'd like for the rest of this chapter. I've mostly been trying to plan the sections I add to this chapter around the order in which I'd like to talk about groups of mitigations.

The next sections I plan to work on after this are code reuse attacks and their mitigations (this will have to be two separate sections, as there is a lot to say!)

After that, the listed skeleton structure currently contains:
"Non-control data exploits"
however I was thinking we can expand that section to lead into more general features for addressing memory vulnerabilities (not only related to code pointers), and in particular I'd also like to describe various sanitisers (UBSan, Address sanitiser etc).

In that case, we need to rethink how to name that section, but overall it seems to me it would possibly be a good place to discuss more general heap buffer overflows and other related issues.

I'd originally planned a separate section to describe hardware support for mitigations, like pointer authentication and memory tagging, but I will probably add an earlier mention of PAuth (at least pac-ret) in code reuse mitigations too.

So, in short, it's all a little bit flexible at the moment. We can, of course, discuss this in more detail in "discussions", as any input would be useful before I get to these later sections, so feel free to start a thread!

File: book.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (g-kouv) [reply book.md:402]: Actually, it's when `main` returns that the attacker can control where we return (since the saved LR is overwritten in main's frame). More of a reason to add a detailed explanation!
